### PR TITLE
Deprecate the special SYMFONY__ environment variables

### PR DESCRIPTION
**Pull request by javiereguiluz**
| Q | A |
| --- | --- |
| Branch? | master |
| Bug fix? | no |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | yes |
| Tests pass? | yes |
| Fixed tickets | #20089 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by chalasr** [comment]
I surely miss something but https://github.com/symfony/symfony/pull/20100#issuecomment-275523325:

>  shouldn't the Kernel::getEnvParameters() method be deprecated?

Or do we plan to trigger this deprecation forever?

**Comment by robfrawley** [comment]
Vastly prefer the implementation described in https://github.com/symfony/symfony/pull/20100#issuecomment-275523325. I was going to comment on this earlier, but couldn't find that comment by @chalasr.

**Comment by javiereguiluz** [comment]
@chalasr about your comment, I don't agree on reporting the deprecation when using the env var instead of when setting it (if you set it, it's because you have defined it ... so you should get the deprecation).

About the other proposal ... I don't know what to say ... but others think it's much better, so if you can implement it, I guess we can close this PR in favor of your PR. Thanks!

**Comment by robfrawley** [comment]
The reason I believe we shouldn't get *any* deprecations until it an environment variable is *actually used* is because otherwise, we are restricting the use of *external* environment variables without knowing that they are intended for consumption by Symfony itself. People may well be using environment variables beginning with `SYMFONY__` for their own container port mapping or other functionality related to setting up their Symfony environment, but external to the Symfony runtime itself.

**Comment by chalasr** [comment]
@javiereguiluz My point is mainly what @robfrawley said in the previous comment, but also that deprecating `SYMFONY__` prefixed env vars does not make sense to me in term of upgrade. 
Defining these prefixed env vars will not stop to work, only using them as parameters will. 
The symfony-specific behavior is only to consider these vars differently, I think the deprecation should reflect this and that this behavior should really die in 4.0 (i.e. `Kernel::getEnvParameters()` be removed).
I'll propose the alternative PR.

**Comment by javiereguiluz** [comment]
@chalasr yes, I agree that both of you are right and your solution is better than mine :) Thanks for providing the alternative PR. I'll close this one then.

**Comment by nicolas-grekas** [comment]
In think this implem here is right, but misses the deprecation of `Kernel::getEnvParameters`
The false positive that has been identified makes no harm at all: it's mainly theoretical, and it's not like it's going to break anything (4.0 won't throw - it'll just ignore and not warn anymore).

The alternative proposed in #21997 looks worse to me because it breaks SRP.

Until someone finds another solution freed from any of the identified drawbacks, I'm :+1: here (I won't be the one working on this as the identified drawback of this PR not that realistic to me.)

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 1]
```diff
@@ -576,6 +576,7 @@ protected function getEnvParameters()
         $parameters = array();
         foreach ($_SERVER as $key => $value) {
             if (0 === strpos($key, 'SYMFONY__')) {
+                @trigger_error(sprintf('The support of special environment variables that start with SYMFONY__ (such as "%s") is deprecated as of 3.3 and will be removed in 4.0. Use the %env()% syntax instead to get the value of environment variables in configuration files.', $key), E_USER_DEPRECATED);
                 $parameters[strtolower(str_replace('__', '.', substr($key, 9)))] = $value;
```
Comment (chalasr) [inline src/Symfony/Component/HttpKernel/Kernel.php]: `%env()`  should be `%%env()%%`

**Comment by javiereguiluz** [comment]
For the record, I still think that comments from Rob and Robin are right ... but I'll update this PR in case we don't find a viable alternative solution.

**Comment by nicolas-grekas** [comment]
Reopening, this looks like the correct approach to me:

> Defining these prefixed env vars will not stop to work, only using them as parameters will.

`SYMFONY__` is not a common prefix. I'd better consider that we can claim triggering a deprecation notice here. In terms of DX, it's also way more important to me to warn users that if they expect a `SYMFONY__` env var to be taken into account, then that won't happen in 4.0, even in the case they don't actually use it as params.

**Comment by chalasr** [comment]
> even in the case they don't actually use it as params.

If not used as a parameter then there is no reason to expect the env var to be taken into account. I don't understand how this is a better approach regarding DX, but let's move on.

**Comment by nicolas-grekas** [comment]
@chalasr since these env vars are used as *default values*, they can very well be defined by the env just in case - but actually not being used because the param is defined.

File: src/Symfony/Component/HttpKernel/Tests/KernelTest.php [hunk 2]
```diff
@@ -735,6 +735,23 @@ public function testKernelWithoutBundles()
     }
 
     /**
+     * @group legacy
+     * @expectedDeprecation The Symfony\Component\HttpKernel\Kernel::getEnvParameters() method is deprecated as of 3.3 and will be removed in 4.0. Use the %s syntax to get the value of any environment variable from configuration files instead.
+     * @expectedDeprecation The support of special environment variables that start with SYMFONY__ (such as "SYMFONY__FOO__BAR") is deprecated as of 3.3 and will be removed in 4.0. Use the %env()% syntax instead to get the value of environment variables in configuration files.
```
Comment (chalasr) [inline src/Symfony/Component/HttpKernel/Tests/KernelTest.php]: The use of `%env()%` makes the assertion fails due to the use of [`assertStringMatchesFormat`](https://github.com/symfony/symfony/blob/master/src/Symfony/Bridge/PhpUnit/Legacy/SymfonyTestsListenerTrait.php#L258). For now I didn't find better than replacing `%env()%` by `%s` in the expected deprecation to solve this issue

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Tests/KernelTest.php]: `%c` is a bit better :)

File: src/Symfony/Component/HttpKernel/Kernel.php [hunk 3]
```diff
@@ -570,12 +570,19 @@ protected function getKernelParameters()
...
     protected function getEnvParameters()
     {
+        if (0 === func_num_args() || func_get_arg(0)) {
+            @trigger_error(sprintf('The %s() method is deprecated as of 3.3 and will be removed in 4.0. Use the %%env()%% syntax to get the value of any environment variable from configuration files instead.', __METHOD__), E_USER_DEPRECATED);
+        }
+
         $parameters = array();
         foreach ($_SERVER as $key => $value) {
             if (0 === strpos($key, 'SYMFONY__')) {
+                @trigger_error(sprintf('The support of special environment variables that start with SYMFONY__ (such as "%s") is deprecated as of 3.3 and will be removed in 4.0. Use the %%env()%% syntax instead to get the value of environment variables in configuration files.', $key), E_USER_DEPRECATED);
```
Comment (chalasr) [inline src/Symfony/Component/HttpKernel/Kernel.php]: not great but we have to exclude the special env vars which are set from the core: `SYMFONY__REDIS_HOST` is one (https://travis-ci.org/symfony/symfony/jobs/211006453)

Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/Kernel.php]: or not use it anymore :)

**Comment by nicolas-grekas** [comment]
missing entries in CHANGELOG / UPGRADE files

**Comment by fabpot** [comment]
👍

File: UPGRADE-3.3.md [hunk 4]
```diff
@@ -181,6 +181,12 @@ HttpKernel
    which will tell the Kernel to use the response code set on the event's
    response object.
 
+ * The `getEnvParameters()` method has been deprecated and will be removed in 4.0.
```
Comment (nicolas-grekas) [inline UPGRADE-3.3.md]: Kernel::getEnvParameters()

File: UPGRADE-4.0.md [hunk 5]
```diff
@@ -321,6 +321,12 @@ HttpKernel
    which will tell the Kernel to use the response code set on the event's
    response object.
 
+ * The `getEnvParameters()` method has been removed.
```
Comment (nicolas-grekas) [inline UPGRADE-4.0.md]: Kernel::getEnvParameters()

File: src/Symfony/Component/HttpKernel/CHANGELOG.md [hunk 6]
```diff
@@ -7,6 +7,8 @@ CHANGELOG
  * Deprecated `LazyLoadingFragmentHandler::addRendererService()`
  * Added `SessionListener`
  * Added `TestSessionListener`
+ * Deprecated `getEnvParameters`
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpKernel/CHANGELOG.md]: Kernel::getEnvParameters()

**Review by nicolas-grekas — COMMENTED** [review]
A few comments, a rebase, and a test fix, then all good on my side.

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/CachePoolsTest.php [hunk 7]
```diff
@@ -19,12 +19,12 @@ class CachePoolsTest extends WebTestCase
 {
     protected function setUp()
     {
-        $_SERVER['SYMFONY__REDIS_HOST'] = getenv('REDIS_HOST');
+        $_SERVER['REDIS_HOST'] = getenv('REDIS_HOST');
     }
```
Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/CachePoolsTest.php]: seems unnecessary

Comment (javiereguiluz) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/CachePoolsTest.php]: I wasn't sure. Can I remove both setUp() and tearDown() ?

Comment (chalasr) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/CachePoolsTest.php]: I would say yes

**Comment by fabpot** [comment]
Tests are broken.

**Review by chalasr — APPROVED** [review]
👍

**Comment by javiereguiluz** [comment]
The last failing tests seem unrelated, so this is now ready!  A big "thank you" to @chalasr for his great help fixing all the failing tests.

**Review by nicolas-grekas — APPROVED** [review]
:+1:

**Comment by fabpot** [comment]
Thank you @javiereguiluz.

**Pull request closed without merging**